### PR TITLE
replacing db calls with dict lookups

### PR DESCRIPTION
**Pull request by calellowitz**
@esoergel cc: @czue @snopoke @sravfeyn 
While investigating http://manage.dimagi.com/default.asp?245323 , we noticed this small optimization that replaces db calls (13k lookups in the case of the icds domain) with dict lookups when pulling the users list for most reports. I am not sure this will be enough of a speedup to make that report load for all users, but should definitely be a step in the right direction.
🎩 to ethan for the initial spot.

### CODE REVIEW

**Review by esoergel — APPROVED** [review]
Thanks for the fix!

File: corehq/apps/reports/util.py [hunk 1]
```diff
@@ -136,9 +136,8 @@ def _create_temp_user(user_id):
 
         if user_filter[HQUserType.REGISTERED].show:
             # now add all the registered users who never submitted anything
-            for user_id in registered_user_ids:
+            for user_id, user in registered_user_ids.items():
```
Comment (orangejenny) [inline corehq/apps/reports/util.py]: Nitpick: `registered_users` would be better than `registered_user_ids`

Comment (esoergel) [inline corehq/apps/reports/util.py]: I'd go with `registered_users_by_id` to indicate it's a dict.  @millerdev and I have a date to refactor this whole function, it's one of those things that started small, then grew and grew until it does like 3 or 4 unrelated things, is really unperformant, and is hard to reason about.

**Review by czue — APPROVED** [review]
no way, we already had it in the dict? that is amazing.

agree with renaming the variable.

File: corehq/apps/reports/util.py (hunk 1 above)
Comment (snopoke) [inline corehq/apps/reports/util.py]: seems like you could just make this a list comprehension:
```
users.extend(user for user in registered_users_by_id.values() if user.user_id not in submitted_user_ids)
```

**Review by snopoke — APPROVED** [review]

File: corehq/apps/reports/util.py [hunk 2]
```diff
@@ -117,14 +117,14 @@ def _create_temp_user(user_id):
             user_filter = HQUserType.all()
         users = []
         submitted_user_ids = get_all_user_ids_submitted(domain)
-        registered_user_ids = dict([(user.user_id, user) for user in CommCareUser.by_domain(domain)])
+        registered_users_by_id = dict([(user.user_id, user) for user in CommCareUser.by_domain(domain)])
         if include_inactive:
-            registered_user_ids.update(dict([(u.user_id, u) for u in CommCareUser.by_domain(domain, is_active=False)]))
+            registered_users_by_id.update(dict([(u.user_id, u) for u in CommCareUser.by_domain(domain, is_active=False)]))
```
Comment (dimagimon) [inline corehq/apps/reports/util.py]: E501 line too long (122 > 115 characters)

File: corehq/apps/reports/util.py [hunk 3]
```diff
@@ -116,15 +116,15 @@ def _create_temp_user(user_id):
         if not user_filter:
             user_filter = HQUserType.all()
         users = []
-        submitted_user_ids = get_all_user_ids_submitted(domain)
-        registered_user_ids = dict([(user.user_id, user) for user in CommCareUser.by_domain(domain)])
+        submitted_user_ids = set(get_all_user_ids_submitted(domain))
+        registered_users_by_id = dict([(user.user_id, user) for user in CommCareUser.by_domain(domain)])
         if include_inactive:
-            registered_user_ids.update(dict([(u.user_id, u) for u in CommCareUser.by_domain(domain, is_active=False)]))
+            registered_users_by_id.update(dict([(u.user_id, u) for u in CommCareUser.by_domain(domain, is_active=False)]))
```
Comment (dimagimon) [inline corehq/apps/reports/util.py]: E501 line too long (122 > 115 characters)

**Comment by calellowitz** [comment]
updated with name change and list comprehension @orangejenny @czue @snopoke

File: corehq/apps/reports/util.py [hunk 4]
```diff
@@ -116,15 +116,15 @@ def _create_temp_user(user_id):
         if not user_filter:
             user_filter = HQUserType.all()
         users = []
-        submitted_user_ids = get_all_user_ids_submitted(domain)
-        registered_user_ids = dict([(user.user_id, user) for user in CommCareUser.by_domain(domain)])
+        submitted_user_ids = set(get_all_user_ids_submitted(domain))
+        registered_users_by_id = dict([(user.user_id, user) for user in CommCareUser.by_domain(domain)])
```
Comment (calellowitz) [inline corehq/apps/reports/util.py]: this is where the dict is initially created that already had the mobile workers

File: corehq/apps/reports/util.py [hunk 5]
```diff
@@ -136,10 +138,7 @@ def _create_temp_user(user_id):
 
         if user_filter[HQUserType.REGISTERED].show:
             # now add all the registered users who never submitted anything
-            for user_id in registered_user_ids:
-                if user_id not in submitted_user_ids:
-                    user = CommCareUser.get_by_user_id(user_id)
-                    users.append(user)
+            users.extend(user for id, user in registered_users_by_id.items()if id not in submitted_user_ids)
```
Comment (emord) [inline corehq/apps/reports/util.py]: nit: it looks weird to not have a space before `if`

File: corehq/apps/reports/util.py (hunk 4 above)
Comment (esoergel) [inline corehq/apps/reports/util.py]: :+1: on the `set()`

File: corehq/apps/reports/util.py [hunk 6]
```diff
@@ -116,15 +116,17 @@ def _create_temp_user(user_id):
...
         users = []
-        submitted_user_ids = get_all_user_ids_submitted(domain)
-        registered_user_ids = dict([(user.user_id, user) for user in CommCareUser.by_domain(domain)])
+        submitted_user_ids = set(get_all_user_ids_submitted(domain))
+        registered_users_by_id = dict([(user.user_id, user) for user in CommCareUser.by_domain(domain)])
         if include_inactive:
-            registered_user_ids.update(dict([(u.user_id, u) for u in CommCareUser.by_domain(domain, is_active=False)]))
+            registered_users_by_id.update(dict(
+                [(u.user_id, u) for u in CommCareUser.by_domain(domain, is_active=False)]
+            ))
```
Comment (esoergel) [inline corehq/apps/reports/util.py]: nit (fine to punt): 
```python
# this assembles a list, then converts that to a dict:
dict([(u._id, u) for u in users])
# this uses a generator so you never build that list:
dict((u._id, u) for u in users)
# this is my preference overall:
{u._id: u for u in users}
```
My hunch is that the performance differences are marginal compared to the db calls here, but it's still good to think about.

**Pull request merged**